File: python3/binary_tree_level_order_traversal.py
"""
Space   : O(n)
Time    : O(n)
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def getDepth(self, root: TreeNode) -> int:
        if not root:
            return 0

        l = self.getDepth(root.left) + 1
        r = self.getDepth(root.right) + 1

        return max(l, r)

    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []

        dep = self.getDepth(root)
        ans = [[] for i in range(dep)]

        # print(ans)

        stack = [(root, 0)]
        while stack:
            node, level = stack.pop(0)
            ans[level].append(node.val)
            if node.left:
                stack.append((node.left, level + 1))
            if node.right:
                stack.append((node.right, level + 1))

        return ans

File: golang/binary_tree_level_order_traversal.go
package golang

// Space : O(n)
// Time	 : O(n)

import (
	"golang/model"
)

func levelOrder(root *model.TreeNode) [][]int {
	var ans [][]int
	if root == nil {
		return ans
	}

	stack := []*model.TreeNode{root}

	for len(stack) > 0 {
		temp := []*model.TreeNode{}
		level_val := []int{}
		for i := 0; i < len(stack); i++ {
			level_val = append(level_val, stack[i].Val)
			if stack[i].Left != nil {
				temp = append(temp, stack[i].Left)
			}
			if stack[i].Right != nil {
				temp = append(temp, stack[i].Right)
			}
		}

		ans = append(ans, level_val)

		if len(temp) > 0 {
			stack = temp
		} else {
			break
		}
	}

	return ans
}
